Fixes for dataset table aliases

labkey-matthewb opened the pull request:
#### Rationale
<!-- Rationale describing why this pull request is needed, what behavior it's adding/changing/removing, etc. (replace this comment) -->

#### Related Pull Requests
- <!-- list of links to related pull requests (replace this comment) -->

#### Changes
- <!-- list of descriptions of changes that are worth noting (replace this comment) -->

labkey-matthewb committed:
Fixes for dataset table aliases

labkey-matthewb committed:
DatasetUpdateService.TestCase

labkey-jeckels commented on study/src/org/labkey/study/model/DatasetDefinition.java:
```diff
@@ -2749,38 +2752,40 @@ public List<Map<String, Object>> getDatasetRows(User u, Collection<String> lsids
-            if (newKey != null)
-                key = newKey;
-            else if ("_row".equals(key))
+            if ("_row".equals(key))
```
It would be nice to introduce a constant for `_row`. Looks like we have a couple of dozen hard-coded references. I know it's not new in this PR.

labkey-matthewb replied:
Yup

labkey-jeckels commented on study/src/org/labkey/study/query/DatasetUpdateService.java:
```diff
@@ -786,4 +799,103 @@ else if (lsids.length > 1)
+        Container _container = null;
+        StudyImpl _junitStudy = null;
+        StudyManager _manager = StudyManager.getInstance();
+        String longName = "this is a very long name (with punctuation) that raises many questions \"?\" about your database design choices";
```
Nice

labkey-matthewb commented:
Now to figure out what unexpected test chaos this caused.

labkey-matthewb committed:
undo setPropertyURI(null) change
narrow to just "Datasets" column

labkey-matthewb committed:
storagecolumnname is varchar(100)

labkey-matthewb committed:
Tests expects participantid to always be logged.
Also, AuditHandler should handle column.name keyed maps.

labkey-matthewb committed:
isLegalNameChar(char ch, boolean first, @Nullable SqlDialect dialect)
Move back to preferring column.getName() to column.getMetaDataName()
finally switch DatasetDefinition.getDatasetRows() to use queryTableInfo
Fixes for column that support missing values

labkey-matthewb committed:
missed one

labkey-matthewb committed:
No leading digits for legalName

labkey-matthewb committed:
No leading digits for legalName

labkey-matthewb committed:
return case-insensitive map

labkey-matthewb committed:
update shouldQuoteIdentifier()

labkey-jeckels commented on api/src/org/labkey/api/audit/AuditHandler.java:
```diff
@@ -64,12 +64,15 @@ static Pair<Map<String, Object>, Map<String, Object>> getOldAndNewRecordForMerge
-                    .map((ColumnInfo::getName))
-                    .findFirst()
-                    .orElse(key);
+            // getDatasetRows() (at least) should returns key==column.getName(), expect getColumn(name) to work
```
```suggestion
            // getDatasetRows() (at least) should return key==column.getName(), expect getColumn(name) to work
```

(at least) compared with what alternative?

labkey-jeckels commented on api/src/org/labkey/api/query/AliasManager.java:
```diff
@@ -60,27 +60,43 @@ public AliasManager(@NotNull TableInfo table, @Nullable Collection<ColumnInfo> c
+        return isLegalName(str, null);
+    }
+
+    public static boolean isLegalName(String str, SqlDialect dialect)
```
```suggestion
    public static boolean isLegalName(String str, @Nullable SqlDialect dialect)

```

labkey-matthewb replied:
This won't be nullable for long.

labkey-jeckels commented on api/src/org/labkey/api/query/AliasManager.java:
```diff
@@ -60,27 +60,43 @@ public AliasManager(@NotNull TableInfo table, @Nullable Collection<ColumnInfo> c
+        return legalNameFromName(str, null);
+    }
+
+    public static String legalNameFromName(String str, SqlDialect dialect)
```
```suggestion
    public static String legalNameFromName(String str, @Nullable SqlDialect dialect)

```

labkey-jeckels commented on api/src/org/labkey/api/query/AliasManager.java:
```diff
@@ -60,27 +60,43 @@ public AliasManager(@NotNull TableInfo table, @Nullable Collection<ColumnInfo> c
+            return true;
+        if (null == dialect || dialect.isOracle())
+            return !first && ch == '_';
+        return ch == '_' || Character.isAlphabetic(ch);
```
I'm guessing this is supporting characters not in the A-Z range? Maybe worth a comment on how this is different than the check at the beginning of this method.

labkey-matthewb replied:
This whole mess is because oracle doesn't allow leading underscore.  Since we allow arbitrary column names in postgres/sqlserver there's no need to be overly prescriptive.

labkey-matthewb replied:
PS, this is check was creating a lot of X__ prefixes because it was disallowing leading _ for fields in our own module tables.

labkey-matthewb replied:
I'll move the isAlphabetic() change to develop.

labkey-jeckels commented on study/src/org/labkey/study/model/DatasetDefinition.java:
```diff
@@ -2702,86 +2705,72 @@ public Map<String, Object> getDatasetRow(User u, String lsid)
-        {
-            keyNames.put(col.getName(), col.getName());
+            // Results gives us key=ColumnInfo.getFieldKey(), but we need key=ColumnInfo.getName()
+            // there are many ways to do this (we could use ResultSet and col.getValue(rs))
```
Am I missing the call to `col.getValue()`?

labkey-matthewb replied:
This is a Results not a ResultSet.  So result columns are associated with their ColumnInfo.  So rs.getFieldKeyRowMap() is doing the heavy lifting here.

labkey-jeckels replied:
Nevermind, I totally misread the comment, apparently. I thought it said `we use ResultSet and co.getValue(rs)`

labkey-jeckels commented on study/src/org/labkey/study/query/DatasetUpdateService.java:
```diff
@@ -786,4 +807,192 @@ else if (lsids.length > 1)
+            assertEquals("lUpdated", map.get(longName));
+            assertNull(map.get("value1"));        // NA
+            assertEquals("NA", map.get("Value1MVIndicator"));
+            assertEquals(2.0d, map.get("value2"));         // 2.0
```
Indentation is odd on all of these comments

labkey-matthewb replied:
Hmmm, tabs messing with me?

labkey-jeckels commented on study/src/org/labkey/study/query/DatasetUpdateService.java:
```diff
@@ -786,4 +807,192 @@ else if (lsids.length > 1)
+        String longName = "this is a very long name (with punctuation) that raises many questions \"?\" about your database design choices";
+
+        @Test
+        public void updateRowTest() throws Exception
```
A good place for a comment and pointer to the issue that provoked this coverage

jryurkanin commented:
are we good to merge this? WNPRC updated the priority on this and wants to know when it'll ready.

labkey-matthewb committed:
Update api/src/org/labkey/api/audit/AuditHandler.java

Co-authored-by: Josh Eckels <jeckels@labkey.com>

labkey-matthewb committed:
Update api/src/org/labkey/api/query/AliasManager.java

Co-authored-by: Josh Eckels <jeckels@labkey.com>

labkey-matthewb committed:
Update api/src/org/labkey/api/query/AliasManager.java

Co-authored-by: Josh Eckels <jeckels@labkey.com>

labkey-matthewb committed:
save .isAlphabetic() change for develop

labkey-matthewb committed:
some spacing

labkey-adam commented:
> are we good to merge this? WNPRC updated the priority on this and wants to know when it'll ready.

@jryurkanin No, but it's being worked on. This is a very sensitive area of the code, so we need to be careful, especially for a 24.11 fix. We will merge it once it's been manually tested to our satisfaction.

labkey-martyp reviewed (approved):
Approving for fixing the original support ticket and the additional repro in the study test.

labkey-matthewb committed:
Stop trying to match ptid->"DataSets"

labkey-matthewb committed:
Merge remote-tracking branch 'origin/release24.11-SNAPSHOT' into 24.11_fb_dataset_aliases
